app/services/work_packages/shared/working_days.rb

```diff
@@ -111,6 +110,26 @@ def assert_strictly_positive_duration(duration)
         raise ArgumentError, "duration must be strictly positive" if duration.is_a?(Integer) && duration <= 0
       end
 
+      def add_lag(date, lag)
+        date_with_lag = date.next_day
+        while lag != 0
+          lag -= 1 if working?(date_with_lag)
+          date_with_lag += 1
+        end
+        date_with_lag
+      end
+
+      def remove_lag(date, lag)
+        date_with_lag = date
+        while lag != 0
+          lag -= 1 if working?(date_with_lag)
+          next if lag == 0
+
+          date_with_lag -= 1
+        end
+        date_with_lag
```

Comment:
I had to solve a similar issue (working day calculations) at a previous job and accomplished it using an Enumerator. Feel free to use or not use this, but I personally find the functional style more readable vs. imperative.

```suggestion
        to_drop = [0, lag - 1].max
        Enumerator.produce(date, &:prev_day)
          .lazy
          .select { working?(it) }
          .drop(to_drop)
          .take(1)
          .first
```

I also tried to come up with a similar implementation for `add_tag` but struggled to get it working [with the above-mentioned behaviour](https://github.com/opf/openproject/pull/19033/files#r2109924606):

```ruby
      def add_lag(date, lag)
        to_drop = [0, lag].max
        Enumerator.produce(date, &:next_day)
          .lazy
          .drop(1)
          .select { working?(it) }
          .drop(to_drop)
          .take(1)
          .first
      end
```

Reply:
Maybe I'm not used to it, but I actually find it harder to read. It's ok until `select { working?(it) }`, but then `drop([0, lag - 1].max)`, and `take(1)` and `first`. I don't know, maybe not my way of seeing the problem.

And for debugging, chained calls are a chore too. Putting a `binding.irb` inside the `while` loop made it easier to debug in my early implementations.

Sorry, I'll keep it like this.

Reply:
No worries. The original is also not not clear.